### PR TITLE
Add random actor placement; fix cancellation callback; update test skips

### LINKED ISSUE #9483: [New scheduler] Pass test_actor*
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Port the new random actor placement PR; enable a few more test suites with fine-grained skips.
Also fix an issue where we didn't set the cancelled bit on task lease returns.

Fixes https://github.com/ray-project/ray/issues/9483

### CODE REVIEW

File: src/ray/raylet/scheduling/cluster_resource_data.cc [hunk 1]
```diff
@@ -311,6 +311,20 @@ TaskResourceInstances NodeResourceInstances::GetAvailableResourceInstances() {
   return task_resources;
 };
 
+bool TaskRequest::IsEmpty() const {
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:314]: This and other new code bits are copied from the previous random actor placement PR.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc [hunk 2]
```diff
@@ -167,6 +168,25 @@ int64_t ClusterResourceScheduler::GetBestSchedulableNode(const TaskRequest &task
...
+      int idx = std::rand() % nodes_.size();
+      for (auto &node : nodes_) {
+        if (idx == 0) {
+          best_node = node.first;
+          break;
+        }
+        idx--;
+      }
+    }
+    RAY_LOG(WARNING) << "GetBestSchedulableNode, best_node = " << best_node
```
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:184]: This should probably be debug level or removed?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 3]
```diff
@@ -178,12 +178,20 @@ void ClusterTaskManager::HandleTaskFinished(std::shared_ptr<WorkerInterface> wor
...
+  callback();
+}
+
 bool ClusterTaskManager::CancelTask(const TaskID &task_id) {
   for (auto shapes_it = tasks_to_schedule_.begin(); shapes_it != tasks_to_schedule_.end();
        shapes_it++) {
     auto &work_queue = shapes_it->second;
     for (auto work_it = work_queue.begin(); work_it != work_queue.end(); work_it++) {
       if (std::get<0>(*work_it).GetTaskSpecification().TaskId() == task_id) {
+        ReplyCancelled(*work_it);
```
Comment (wuisawesome) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:194]: Can we assert this in `TaskCancellationTest`?

File: src/ray/raylet/worker_pool.cc [hunk 4]
```diff
@@ -1036,7 +1036,7 @@ std::string WorkerPool::DebugString() const {
     result << "\n- num " << Language_Name(entry.first)
            << " drivers: " << entry.second.registered_drivers.size();
   }
-  result << "- num idle workers: " << idle_of_all_languages_.size();
+  result << "\n- num idle workers: " << idle_of_all_languages_.size();
```
Comment (wuisawesome) [inline src/ray/raylet/worker_pool.cc:1039]: is this intentional?

Comment (ericl) [reply src/ray/raylet/worker_pool.cc:1039]: Yeah it was missing a newline before.

**Comment by ericl** [comment]
@wuisawesome updated

**Review by wuisawesome — APPROVED** [review]

**Pull request merged**